linux/LinuxProcessList.c

```diff
@@ -629,45 +623,85 @@ static void LinuxProcessList_readMaps(LinuxProcess* process, openat_arg_t procFd
       if (!map_inode)
          continue;
 
-      if (calcSize) {
-         LibraryData* libdata = Hashtable_get(ht, map_inode);
-         if (!libdata) {
-            libdata = xCalloc(1, sizeof(LibraryData));
-            Hashtable_put(ht, map_inode, libdata);
-         }
-
-         libdata->size += map_end - map_start;
-         libdata->exec |= map_execute;
+      LibraryData* libdata = Hashtable_get(ht, map_inode);
+      if (!libdata) {
+         libdata = xCalloc(1, sizeof(LibraryData));
+         Hashtable_put(ht, map_inode, libdata);
       }
 
-      if (checkDeletedLib && map_execute && !proc->usesDeletedLib) {
-         while (*readptr == ' ')
-            readptr++;
+      libdata->size += map_end - map_start;
+      libdata->exec |= map_execute;
+   }
 
-         if (*readptr != '/')
-            continue;
+   fclose(mapsfile);
 
-         if (String_startsWith(readptr, "/memfd:"))
-            continue;
+   uint64_t total_size = 0;
+   Hashtable_foreach(ht, LinuxProcessList_calcLibSize_helper, &total_size);
 
-         if (strstr(readptr, " (deleted)\n")) {
-            proc->usesDeletedLib = true;
-            if (!calcSize)
-               break;
-         }
-      }
+   Hashtable_delete(ht);
+
+   process->m_lrs = total_size / pageSize;
+}
+
+static void LinuxProcessList_readMapFiles(LinuxProcess* lp, openat_arg_t procFd) {
+   Process* proc = (Process*)lp;
+   proc->usesDeletedLib = false;
+
+#ifndef HAVE_OPENAT
+
+   /*
+    * /proc/[pid]/map_files/ exists since kernel 3.3;
+    * use openat(2), available since kernel 2.6.16, unconditionally.
+    */
+   (void) procFd;
+   return;
+
+#else /* !HAVE_OPENAT */
+
+   int fd = openat(procFd, "map_files", O_RDONLY | O_DIRECTORY | O_NOFOLLOW);
+   if (fd < 0)
+      return;
+
+   DIR* dir = fdopendir(fd);
+   if (!dir) {
+      close(fd);
+      return;
    }
 
-   fclose(mapsfile);
+   const struct dirent* entry;
+   while ((entry = readdir(dir)) != NULL) {
+      /* Ignore all non-symlinks */
+      if (entry->d_type != DT_LNK && entry->d_type != DT_UNKNOWN)
+         continue;
+
+      char linkBuf[PATH_MAX];
+      ssize_t rc = readlinkat(fd, entry->d_name, linkBuf, sizeof(linkBuf));
+      if (rc < 0)
+         continue;
 
-   if (calcSize) {
-      uint64_t total_size = 0;
-      Hashtable_foreach(ht, LinuxProcessList_calcLibSize_helper, &total_size);
+      linkBuf[rc] = '\0';
 
-      Hashtable_delete(ht);
+      if (String_startsWith(linkBuf, "/memfd:"))
```

Comment:
Also ignore `/dev/shm/` (Shared memory segments) causing false positives for e.g. chromium.